### PR TITLE
BUG: pivot_table dropna=False drops columns/index names

### LINKED ISSUE #12133: crosstab dropna=False breaks columns.names
Specifying `dropna=False` drops column level names:
```
import pandas as pd
df = pd.DataFrame({"Col1": [23], "Col2": [45]})
names1 = pd.crosstab(df["Col1"], df["Col2"]).columns.names
names2 = pd.crosstab(df["Col1"], df["Col2"], dropna=False).columns.names
print(names1, names2)
>>>['Col2'] [None]
```
P.S. At first i thought `dropna=False` will help to count `NaN` values, though it seems that `fillna` is needed. ~~So i don't understand, why this parameter is needed at all.~~ If entire column is `NaN` then we'll get `KeyError: '__dummy__'` anyway.

pandas: 0.17.1

### LINKED ISSUE #12642: BUG: crosstab with margins=True and dropna=False
xref #12614 
```
In [1]: df = pd.DataFrame({'a': [1, 2, 2, 2, 2, np.nan],'b': [3, 3, 4, 4, 4, 4]})

In [2]: pd.crosstab(df.a, df.b, margins=True, dropna=True)
Out[2]: 
b    3  4  All
a             
1.0  1  0    1
2.0  1  3    4
All  2  3    5

In [3]: pd.crosstab(df.a, df.b, margins=True, dropna=False)
KeyError: 'Level None not found'
```

### PR DESCRIPTION
**Pull request by OXPHOS**
- [X] closes #12133 
- [X] closes #12642, closes #12327 (replace PR)
- [X] tests added / passed
- [X] passes `git diff upstream/master | flake8 --diff`
- [X] whatsnew entry

Bug fixed from pull request #12327.

### CODE REVIEW

File: pandas/tools/tests/test_pivot.py [hunk 1]
```diff
@@ -965,6 +965,53 @@ def test_margin_ignore_dropna_bug(self):
...
+        tm.assert_frame_equal(actual, expected)
+
+        a = np.array(['foo', 'foo', 'foo', 'bar',
+                      'bar', 'foo', 'foo'], dtype=object)
+        b = np.array(['one', 'one', 'two', 'one',
+                      'two', np.nan, 'two'], dtype=object)
+        c = np.array(['dull', 'dull', 'dull', 'dull',
+                      'dull', 'shiny', 'shiny'], dtype=object)
+
+        res = crosstab(a, [b, c], rownames=['a'],
```
Comment (jreback) [inline pandas/tools/tests/test_pivot.py]: pd.crosstab (and same below)

File: pandas/tools/tests/test_pivot.py [hunk 2]
```diff
@@ -965,6 +965,53 @@ def test_margin_ignore_dropna_bug(self):
...
+                       colnames=['b', 'c'], margins=True, dropna=False)
+        m = MultiIndex.from_tuples([('one', 'dull'), ('one', 'shiny'),
+                                    ('two', 'dull'), ('two', 'shiny'),
+                                    ('All', '')])
+        assert_equal(res.columns.values, m.values)
```
Comment (jreback) [inline pandas/tools/tests/test_pivot.py]: you need to contruct an expected frame and use ``assert_frame_equal``

**Comment by OXPHOS** [comment]
@jreback Thanks for the comments. But I don't think this is the correct way to cherry pick commits..?? I'll learn some github basics and fix this thing within next two days.

**Comment by jreback** [comment]
@OXPHOS it looks ok. you prob want to squash commits down though to make it easier to work with (e.g. squash yours to a single one, his the same).

**Comment by OXPHOS** [comment]
Cool. Got it. Thanks!

**Comment by OXPHOS** [comment]
Hey I have tried my best to squash the commits and I guess I screwed up something in the very beginning... This is the best result I can get so far. Can you please check the conflict if possible? Sorry for the inconvenience.

**Comment by OXPHOS** [comment]
@jreback ready to merge : )

File: pandas/tools/tests/test_pivot.py [hunk 3]
```diff
@@ -965,6 +965,62 @@ def test_margin_ignore_dropna_bug(self):
         expected.columns = Index([3, 4, 'All'], name='b')
         tm.assert_frame_equal(actual, expected)
 
+        # GH 12642
```
Comment (jreback) [inline pandas/tools/tests/test_pivot.py]: add this to the whatsnew (as a separate bug fix line)

**Comment by jreback** [comment]
thanks @OXPHOS

**Pull request closed without merging**